### DISCUSSION TITLE
Help ++ Random affine Transformation with more than one Scalar Image in subject++

### DISCUSSION THREAD

**Discussion by nkoser** [TorchIO-project/torchio#759] [Q&A]
Hello dear community, I have the following problem that I have a subject consisting of several scalar images and would like to perform a random affine transformation on this subject. So far this works fine, but the scalar images in the subject have been subjected to a different transformation. How can I do that, that I insert three images (two images and a mask) into a subject in such a way, that the applied transformation is also the same on all three images ? 
Many thanks and greetings 
Niklas

**Comment by fepegar** [comment]
Hi, @nkoser. Could you please share a minimal reproducible example so we can help?

**Comment by nkoser** [comment]
```python
class torchIO_RandomAffine(object):

    def __init__(self, scales: TypeOneToSixFloat = 0.1,
                 degrees: TypeOneToSixFloat = 10,
                 translation: TypeOneToSixFloat = 0,
                 isotropic: bool = False,
                 center: str = 'image',
                 default_pad_value='minimum',
                 image_interpolation: str = 'linear', ):
        self.scales = scales
        self.degrees = degrees
        self.translation = translation
        self.isotropic = isotropic
        self.center = center
        self.default_pad_value = default_pad_value
        self.image_interpolation = image_interpolation

    def __call__(self, sample):
        org, sub, seg = sample
        subject = tio.Subject(
            original=tio.ScalarImage(tensor=org),
            subsampled=tio.ScalarImage(tensor=sub),
            mask=tio.LabelMap(tensor=seg))
        affine = tio.RandomAffine(scales=self.scales,
                                  degrees=self.degrees,
                                  translation=self.translation,
                                  isotropic=self.isotropic,
                                  center=self.center,
                                  default_pad_value=self.default_pad_value,
                                  image_interpolation=self.image_interpolation)
        subj = affine(subject)
        # print(subj.get_images())
        return subj.subsampled.data, subj.original.data, subj.mask.data`
```


This is the small part I have so far. I have now extra not made in a record, because that was just too much effort for me, as I can load .mhd and raw files with you. Anyway my two patches org and sub look like this
![test_0](https://user-images.githubusercontent.com/58036930/143622092-35ea730d-687e-4214-bdf4-7b4d8c0eaf76.png)
![test_1](https://user-images.githubusercontent.com/58036930/143622094-45c2d00d-c038-4834-b11c-76f1e08a1306.png)

**Reply by fepegar** [reply]
Thanks. This is not a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example): I cannot run that on my computer and get the same results. Could you please write one and share some data to reproduce it?

**Reply by nkoser** [reply]
Oh sorry, sure, just a moment. In any case, thank you very much for your help

**Reply by nkoser** [reply]
I have now built the following code snippet and somehow this does not show up. Possibly I have done something wrong. I wanted to ask also; 
-  whether there is a possibility from the Subject.ScalarImage.data to get the affine transformation of the Scalar Image or is it only possible to get it from the subject ? 
- Is it possible to run these preprocessing steps on the graphic cards ?

**Reply by nkoser** [reply]
```python
import torch
import torchio as tio
from torchio.transforms.augmentation.spatial.random_affine import TypeOneToSixFloat


class torchIO_RandomAffine(object):

    def __init__(self, scales: TypeOneToSixFloat = 0.1,
                 degrees: TypeOneToSixFloat = 10,
                 translation: TypeOneToSixFloat = 0,
                 isotropic: bool = False,
                 center: str = 'image',
                 default_pad_value='minimum',
                 image_interpolation: str = 'linear', ):
        self.scales = scales
        self.degrees = degrees
        self.translation = translation
        self.isotropic = isotropic
        self.center = center
        self.default_pad_value = default_pad_value
        self.image_interpolation = image_interpolation

    def __call__(self, sample):
        org, sub, seg = sample
        subject = tio.Subject(
            original=tio.ScalarImage(tensor=org),
            subsampled=tio.ScalarImage(tensor=sub),
            mask=tio.LabelMap(tensor=seg))
        affine = tio.RandomAffine(scales=self.scales,
                                  degrees=self.degrees,
                                  translation=self.translation,
                                  isotropic=self.isotropic,
                                  center=self.center,
                                  default_pad_value=self.default_pad_value,
                                  image_interpolation=self.image_interpolation)
        subj = affine(subject)
        # print(subj.get_images())
        return subj.subsampled.data, subj.original.data, subj.mask.data


t = torchIO_RandomAffine()
d = torch.load('ABD_LYMPH_0010org.pt')
org = d['data'].cpu()
seg = d['seg'].cpu()
d = torch.load('ABD_LYMPH_0010sub.pt')
sub = d['data'].cpu()
print(org.shape)
print(sub.shape)
print(seg.shape)
org, sub, seg = t((org, sub, seg.squeeze(0)))


from matplotlib import pyplot as plt

plt.imshow(org[0,:,32,:].cpu(),cmap='gray')
plt.imshow(sub[0,:,32,:],cmap='gray')
```

**Reply by nkoser** [reply]
> I have now built the following code snippet and somehow this does not show up. Possibly I have done something wrong. I wanted to ask also;
> 
> * whether there is a possibility from the Subject.ScalarImage.data to get the affine transformation of the Scalar Image or is it only possible to get it from the subject ?
> * Is it possible to run these preprocessing steps on the graphic cards ?

I have problem to upload some .pt data, so I will set up a repository where you can access the data. Thank you

**Reply by nkoser** [reply]
https://github.com/nkoser/data.git

**Reply by fepegar** [reply]
Can you please share non-serialized data (e.g., NIfTI)? [I'd rather not have to use `torch.load`](https://github.com/pytorch/pytorch/issues/52596#issuecomment-783580396):
![](https://user-images.githubusercontent.com/98330/108752433-bb443480-7543-11eb-9e17-1a9149e64ab3.png)
